netmon/netmon.go

```diff
@@ -0,0 +1,613 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log/syslog"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	lSyslog "github.com/sirupsen/logrus/hooks/syslog"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+// The following types and structures have to be kept in sync with the
+// description of the agent protocol. Those definitions need to be in their
+// own separate package so that they can be imported directly from this code.
+// The reason for not importing them now, is because importing the whole agent
+// protocol adds up too much overhead because of the grpc protocol involved.
```

Comment:
No, the real solution is to move the types out of this protocol package. They need to be in their own package so that we can only import the structure definition. But that's definitely something that needs to be done in parallel, and I didn't want to delay this PR because moving things out might take some time.

Reply:
Oh - I see what you mean. Something like this?

```diff
diff --git a/protocols/grpc/agent.proto b/protocols/grpc/agent.proto
index 4c09d9c..f7f27d4 100644
--- a/protocols/grpc/agent.proto
+++ b/protocols/grpc/agent.proto
@@ -10,6 +10,7 @@ package grpc;
 
 import "oci.proto";
 import "google/protobuf/empty.proto";
+import "types.proto";
 
 
 // unstable
@@ -270,42 +271,10 @@ message CreateSandboxRequest {
 message DestroySandboxRequest {
 }
 
-enum IPFamily {
-	v4 = 0;
-	v6 = 1;
-}
-
-message IPAddress {
-	IPFamily family = 1;
-	string address = 2;
-	string mask = 3;
-}
-
-message Interface {
-	string device = 1;
-	string name = 2;
-	repeated IPAddress IPAddresses = 3;
-	uint64 mtu = 4;
-	string hwAddr = 5;
-
-	// pciAddr is the PCI address in the format  "bridgeAddr/deviceAddr".
-	// Here, bridgeAddr is the address at which the bridge is attached on the root bus,
-	// while deviceAddr is the address at which the network device is attached on the bridge.
-	string pciAddr = 6;
-}
-
 message Interfaces {
 	repeated Interface Interfaces = 1;
 }
 
-message Route {
-	string dest = 1;
-	string gateway = 2;
-	string device = 3;
-	string source = 4;
-	uint32 scope = 5;
-}
-
 message Routes {
 	repeated Route Routes = 1;
 }
diff --git a/protocols/grpc/types.proto b/protocols/grpc/types.proto
new file mode 100644
index 0000000..823022f
--- /dev/null
+++ b/protocols/grpc/types.proto
@@ -0,0 +1,35 @@
+syntax = "proto3";
+
+package grpc;
+
+enum IPFamily {
+	v4 = 0;
+	v6 = 1;
+}
+
+message IPAddress {
+	IPFamily family = 1;
+	string address = 2;
+	string mask = 3;
+}
+
+message Interface {
+	string device = 1;
+	string name = 2;
+	repeated IPAddress IPAddresses = 3;
+	uint64 mtu = 4;
+	string hwAddr = 5;
+
+	// pciAddr is the PCI address in the format  "bridgeAddr/deviceAddr".
+	// Here, bridgeAddr is the address at which the bridge is attached on the root bus,
+	// while deviceAddr is the address at which the network device is attached on the bridge.
+	string pciAddr = 6;
+}
+
+message Route {
+	string dest = 1;
+	string gateway = 2;
+	string device = 3;
+	string source = 4;
+	uint32 scope = 5;
+}
```

That seems to work, so why can't we JFDI? :)

Reply:
Doh - as @sboeuf has pointed out, although this "works", it doesn't actually do what we want - we want just the types in a `types.go`, we don't want all the gobs of auto-generated functions that grpc puts into `types.pb.go`.